In [49]:
# Import pandas and numpy
import pandas as pd
import numpy as np
# Import the required dependencies from sklearn
from sklearn.model_selection import train_test_split
from sklearn.pipeline import Pipeline
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC
from sklearn import metrics
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.svm import LinearSVC


# Set the column width to view the text message data.
pd.set_option('max_colwidth', 200)

In [50]:
# Load the  review dataset.
file_path = 'C:/Users/Admin/Documents/MSU/FINAL - PROJECT/Combined Data.csv'
df = pd.read_csv(file_path)

# Set the column width to view the text message data
pd.set_option('display.max_colwidth', None)

# Display the first five rows of the dataset. 
print(df.head())



   Unnamed: 0  \
0           0   
1           1   
2           2   
3           3   
4           4   

                                                                        statement  \
0                                                                      oh my gosh   
1                trouble sleeping, confused mind, restless heart. All out of tune   
2  All wrong, back off dear, forward doubt. Stay in a restless and restless place   
3                   I've shifted my focus to something else but I'm still worried   
4        I'm restless and restless, it's been a month now, boy. What do you mean?   

    status  
0  Anxiety  
1  Anxiety  
2  Anxiety  
3  Anxiety  
4  Anxiety  


In [51]:
# Display the first five rows of the dataset as a table
print(df.head().to_string(index=False))

# Check the column names in the dataset
print(df.columns)


 Unnamed: 0                                                                      statement  status
          0                                                                     oh my gosh Anxiety
          1               trouble sleeping, confused mind, restless heart. All out of tune Anxiety
          2 All wrong, back off dear, forward doubt. Stay in a restless and restless place Anxiety
          3                  I've shifted my focus to something else but I'm still worried Anxiety
          4       I'm restless and restless, it's been a month now, boy. What do you mean? Anxiety
Index(['Unnamed: 0', 'statement', 'status'], dtype='object')


In [52]:
#Split the data into training and testing sets:
X_train, X_test, y_train, y_test = train_test_split(df[['statement']], df['status'], test_size=0.2, random_state=42)

In [53]:
import numpy as np
# Load the review dataset
file_path = 'C:/Users/Admin/Documents/MSU/FINAL - PROJECT/Combined Data.csv'
df = pd.read_csv(file_path)

df.replace('', np.nan, inplace=True)
df = df.dropna()


In [54]:
# Check if the "Unnamed: 0" column exists
if 'Unnamed: 0' in df.columns:
    # Drop the "Unnamed: 0" column
    df = df.drop('Unnamed: 0', axis=1)

# Display the remaining columns
df.iloc[290:294]


statement  \
290                                                                                                                                           For a long time, my heart feels like it's going to burst when it's restless   
291  Now, he's always anxious every day, it feels like he's waiting for a thesis trial, you know. I understand it feels nervous, worried, afraid. what's the difference now that you don't know what you're worried about   
292                                                                                                                                                                                   The habit of being anxious is clear   
294                                                                                                                                                                                       sleep but restless. understand?   

      status  
290  Anxiety  
291  Anxiety  
292  Anxiety  
294  Anxiety

In [55]:
X = df['statement']
y = df['status']
y.head()



0    Anxiety
1    Anxiety
2    Anxiety
3    Anxiety
4    Anxiety
Name: status, dtype: object

In [56]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


In [57]:
X_train.head()

26386                                                                                                                                                                                                                                                                                                                                I know it seems like such a stupid reason to decide that maybe now is the right time to take this leap but it just feels like confirmation. My friends go out, they celebrate together, their families are proud of them and celebrate with them too. I am celebrating by sitting in my bedroom alone. My dad gave me a hug which was nice and made me feel better for a moment but the knowledge that all of these things I am proud of are pretty much worthless in the end has kind of cemented the idea that it is not really worth being here after all. I do not think my death would really be any different to any of the other unacknowledged milestones in my life. All that any of this 

In [58]:
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer(stop_words=None)),
    ('classifier', LinearSVC())
])

pipeline.fit(X_train, y_train)

# Print the results
predictions = pipeline.predict(X_test)
print(predictions)


['Normal' 'Normal' 'Bipolar' ... 'Suicidal' 'Depression'
 'Personality disorder']


In [59]:
# Make predictions on the test data
predictions = pipeline.predict(X_test)

# Show the test data and corresponding predictions
test_results = pd.DataFrame({'X_test': X_test, 'Predicted_y': predictions, 'Actual_y': y_test})
print(test_results)


In [60]:
# Make predictions on the test data
predictions = pipeline.predict(X_test)

# display prediction 

In [61]:
import pandas as pd
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression

# Preprocessing
df['statement'] = df['statement'].str.lower()  # Convert statements to lowercase

# Feature Extraction
vectorizer = TfidfVectorizer()
X = vectorizer.fit_transform(df['statement'])  # Extract TF-IDF features from the statements

# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, df['status'], test_size=0.2, random_state=42)

# Model Training
model = LogisticRegression()
model.fit(X_train, y_train)

# Model Evaluation
accuracy = model.score(X_test, y_test)
print("Accuracy:", accuracy)

# Prediction
new_statements = ["I feel anxious", "I am happy today"]
new_statements_preprocessed = [statement.lower() for statement in new_statements]
new_statements_features = vectorizer.transform(new_statements_preprocessed)
predictions = model.predict(new_statements_features)
print("Predictions:", predictions)


Accuracy: 0.7691942678181646
Predictions: ['Anxiety' 'Depression']


c:\Users\Admin\anaconda3\envs\dev\lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [62]:
# Replace 'status' with the actual column name for mental health topics
mental_health_topics = df['status'].value_counts()

# Print the total count of the most common mental health topics
print(mental_health_topics)


status
Normal                  16343
Depression              15404
Suicidal                10652
Anxiety                  3841
Bipolar                  2777
Stress                   2587
Personality disorder     1077
Name: count, dtype: int64


In [63]:

# Replace 'status' with the actual column name for mental health topics
X_train = vectorizer.transform(df['statement'])  # Extract TF-IDF features from the training statements

# Fit the model to the training data
model.fit(X_train, df['status'])

# Print the outcome
print("Model fitted to the training data.")


Model fitted to the training data.


c:\Users\Admin\anaconda3\envs\dev\lib\site-packages\sklearn\linear_model\_logistic.py:469: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


In [64]:

# Set the column names for the text data and label/target data
X_column = 'statement'  # Replace 'statement' with the actual column name for the text data
y_column = 'status'  # Replace 'status' with the actual column name for the label/target data

## Split the data into testing data sets.

In [65]:
# Set features variable
features = df['statement']

# Set target variable
target = df['status']

# Split data into training and testing sets
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(features, target, test_size=0.2, random_state=42)


In [66]:
#  Build a pipeline using `TfidfVectorizer()`, without `stopwords='english`, and `LinearSVC()`.
pipeline = Pipeline([
    ('tfidf', TfidfVectorizer(stop_words=None)),
    ('classifier', LinearSVC())
])


# Fit the model to the transformed data.
pipeline.fit(features, target)


Pipeline(steps=[('tfidf', TfidfVectorizer()), ('classifier', LinearSVC())])

In [67]:
# Validate the model by checking the model's training and testing accuracy.
training_accuracy = pipeline.score(X_train, y_train)
testing_accuracy = pipeline.score(X_test, y_test)


## Run predictions and analyze the results.

In [68]:
# Make predictions for the test data
X_test_features = vectorizer.transform(X_test)  # Extract TF-IDF features from the test statements
predictions = model.predict(X_test_features)  # Make predictions for the test data

# Retrieve the first 30 predictions
first_30_predictions = predictions[:30]

# Print the first 30 predictions
print("First 30 predictions:", first_30_predictions)


First 30 predictions: ['Normal' 'Normal' 'Bipolar' 'Depression' 'Normal' 'Suicidal' 'Normal'
 'Normal' 'Depression' 'Depression' 'Suicidal' 'Personality disorder'
 'Depression' 'Depression' 'Depression' 'Depression' 'Depression'
 'Suicidal' 'Normal' 'Anxiety' 'Depression' 'Depression' 'Bipolar'
 'Depression' 'Anxiety' 'Depression' 'Normal' 'Normal' 'Depression'
 'Depression']


In [69]:
from sklearn.metrics import confusion_matrix, classification_report, accuracy_score

# Create the confusion matrix
confusion_mat = confusion_matrix(y_test, predictions)

# Print the confusion matrix
print("Confusion Matrix:")
print(confusion_mat)

# Print a classification report
print("Classification Report:")
print(classification_report(y_test, predictions))

# Print the overall accuracy
accuracy = accuracy_score(y_test, predictions)
print("Overall Accuracy:", accuracy)


Confusion Matrix:
[[ 639    7   48   44    1   10    6]
 [   6  403   75   25    2    8    8]
 [  26   10 2446  127    1   14  392]
 [   8    4   38 3212    1   22   23]
 [   2    2   55   17  136   17    8]
 [  32    7   85   55    1  342   14]
 [   1    3  413  101    0    3 1637]]
Classification Report:
                      precision    recall  f1-score   support

             Anxiety       0.89      0.85      0.87       755
             Bipolar       0.92      0.76      0.84       527
          Depression       0.77      0.81      0.79      3016
              Normal       0.90      0.97      0.93      3308
Personality disorder       0.96      0.57      0.72       237
              Stress       0.82      0.64      0.72       536
            Suicidal       0.78      0.76      0.77      2158

            accuracy                           0.84     10537
           macro avg       0.86      0.77      0.81     10537
        weighted avg       0.84      0.84      0.83     10537

Overall

In [70]:
# Validate the model by checking the model accuracy with model.score
testing_accuracy = pipeline.score(X_test, y_test)


### Feed the previous review into the model's `predict()` method.

In [71]:
# Print the classification of the review.
review_text = "This movie is amazing!"  # Replace with the actual review text
review_features = vectorizer.transform([review_text])
prediction = model.predict(review_features)

print("Review Classification:", prediction)


Review Classification: ['Normal']


-Precision: Measures the proportion of correctly predicted instances for each class. 
Higher precision values indicate fewer false positives. The highest precision is observed for "Personality disorder" (0.96), "Normal" (0.90), and "Bipolar" (0.92).

.Recall: Measures the proportion of actual instances of each class that were correctly predicted. 
Higher recall values indicate fewer false negatives. The highest recall is observed for "Normal" (0.97), "Anxiety" (0.85), and "Depression" (0.81).

- F1-score: The harmonic mean of precision and recall, providing a balanced measure of model performance. 
The highest F1-scores are observed for "Normal" (0.93), "Anxiety" (0.87), and "Bipolar" (0.84).

- Support: Represents the number of instances in each class.

- Accuracy: The overall accuracy of the model is 0.84, indicating the proportion of correctly classified instances out of the total number of instances.

- Macro average F1-score: 0.81, indicating the overall performance of the model across all classes.

- Weighted average F1-score: 0.83, considering the class imbalance.

